File: web/lib/slugs.ts
/**
 * حل‌کننده‌ی اسلاگ تخت — پشت مسیر ‎app/[slug]‎ (بلیت ۷؛ بند ۱۳، تصمیم ۱۱).
 *
 * مالک جدول اسلاگ گردآورنده است (`collector/src/mazane_collector/slugs.py`)؛
 * این ماژول فقط داده‌ی همان جدول را که از استور می‌آید حل می‌کند:
 * اسلاگ دارایی ⟸ صفحه‌ی دارایی، اسلاگ سکو ⟸ صفحه‌ی سکو، ناشناخته ⟸ 404.
 *
 * کلمات رزرو (و صفحات ایستای سطح ریشه) هرگز نباید به این مسیر برسند —
 * مسیرهای ایستای نکست خودشان مقدم‌اند — ولی حل‌کننده به‌هرحال ردشان می‌کند
 * (دفاع در عمق: حتی اگر روزی payload آلوده اسلاگ «blog» را ادعا کند، این
 * مسیر آن را نمی‌گیرد). فهرست، آینه‌ی RESERVED_WORDS گردآورنده است.
 *
 * دروازه‌ی انتشار (تصمیم ۱۰) اینجا فقط **خوانده** می‌شود: دارایی با
 * published=false مثل ناشناخته 404 است. آستانه و شمارش در گردآورنده است.
 */
import {
  getInstruments,
  getListedPlatforms,
  type InstrumentListing,
  type ListedPlatform,
} from "./prices";

/** آینه‌ی RESERVED_WORDS جدول مرکزی گردآورنده (بند ۱۳، تصمیم ۱۱). */
export const RESERVED_WORDS: ReadonlySet<string> = new Set([
  "blog",
  "go",
  "api",
  "sitemap.xml",
  "robots.txt",
  "_next",
  "about",
]);

/** صفحات ایستای سطح ریشه — مسیر ایستای خودشان را دارند، نه ‎[slug]‎. */
export const STATIC_PAGE_SLUGS: ReadonlySet<string> = new Set([
  "darbare-pishnahad",
  "mazane-chist",
]);

export type SlugResolution =
  | { kind: "instrument"; listing: InstrumentListing }
  | { kind: "platform"; platform: ListedPlatform };

/** رزرو یا صفحه‌ی ایستا — هرگز از مسیر داینامیک حل نمی‌شود. */
export function isReservedSlug(slug: string): boolean {
  return RESERVED_WORDS.has(slug) || STATIC_PAGE_SLUGS.has(slug);
}

/**
 * حل اسلاگ به موجودیت — null یعنی 404 (ناشناخته، رزروشده، یا دارایی‌ای که
 * دروازه‌ی انتشارش بسته است).
 */
export async function resolveSlug(slug: string): Promise<SlugResolution | null> {
  if (isReservedSlug(slug)) return null;
  const [instruments, platforms] = await Promise.all([
    getInstruments(),
    getListedPlatforms(),
  ]);
  const listing = instruments.find((item) => item.slug === slug);
  if (listing !== undefined) {
    // دارایی تک‌سکویی صفحه نمی‌گیرد (تصمیم ۱۰) — پرچم گردآورنده، نه شمارش وب.
    return listing.published ? { kind: "instrument", listing } : null;
  }
  const platform = platforms.find((item) => item.slug === slug);
  if (platform !== undefined) return { kind: "platform", platform };
  return null;
}

File: collector/src/tablo_collector/slugs.py
from __future__ import annotations

import re
from enum import StrEnum

from .instruments import INSTRUMENTS
from .platforms import PLATFORMS

RESERVED_WORDS = frozenset(
    {"blog", "go", "api", "sitemap.xml", "robots.txt", "_next", "about"}
)

STATIC_PAGE_SLUGS: tuple[str, ...] = (
    "mazane-chist",
    "sekeh",
    "methodology",
    "mohasebe-tala",
    "mohasebe-forush-tala",
    "abzarha",
)

_SLUG_PATTERN = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


class SlugKind(StrEnum):
    PLATFORM = "platform"
    INSTRUMENT = "instrument"
    STATIC_PAGE = "static_page"
    BLOG_POST = "blog_post"


class SlugError(ValueError):
    pass


class InvalidSlugError(SlugError):
    pass


class ReservedSlugError(SlugError):
    pass


class SlugCollisionError(SlugError):
    pass


class SlugRegistry:

    def __init__(self) -> None:
        self._slugs: dict[str, SlugKind] = {}

    def register(self, slug: str, kind: SlugKind) -> None:
        self.validate_new_slug(slug)
        self._slugs[slug] = kind

    def validate_new_slug(self, slug: str) -> None:
        if slug in RESERVED_WORDS:
            raise ReservedSlugError(f"slug {slug!r} is a reserved word (decision 11)")
        if not _SLUG_PATTERN.fullmatch(slug):
            raise InvalidSlugError(
                f"slug {slug!r} is not flat Latin (only a-z0-9 and hyphen — clause 6.6)"
            )
        if slug in self._slugs:
            raise SlugCollisionError(
                f"slug {slug!r} was already given to {self._slugs[slug].value} — uniqueness constraint"
            )

    def kind_of(self, slug: str) -> SlugKind | None:
        return self._slugs.get(slug)

    def __contains__(self, slug: str) -> bool:
        return slug in self._slugs


def build_registry() -> SlugRegistry:
    registry = SlugRegistry()
    for platform in PLATFORMS:
        registry.register(platform.slug, SlugKind.PLATFORM)
    for info in INSTRUMENTS:
        registry.register(info.slug, SlugKind.INSTRUMENT)
    for slug in STATIC_PAGE_SLUGS:
        registry.register(slug, SlugKind.STATIC_PAGE)
    return registry


PUBLIC_SLUGS = build_registry()
